feat(crypto): implement encrypted signing key storage

LeKarimDerradji opened the pull request:
Implement passphrase-protected encryption/decryption for Ed25519 signing keys.

**Implementation:**
- Add write_signing_key_to_file() - encrypts and saves key to disk
- Add read_signing_key_from_file() - loads and decrypts key from disk
- Add derive_key_from_passphrase() - Argon2 KDF helper

**Cryptography:**
- Argon2 (default params) for key derivation from passphrase
- ChaCha20Poly1305 AEAD for authenticated encryption
- File format: 76 bytes [salt:16][nonce:12][ciphertext+tag:48]

**Tests:**
- Round-trip encryption/decryption
- Wrong passphrase rejection
- File corruption detection
- Invalid file size handling
- Argon2 consistency

**Dependencies:**
- argon2 = "0.5.3"
- chacha20poly1305 = "0.10.1"
- hex = "0.4" (for tests)

Closes #3

LeKarimDerradji committed:
feat(crypto): implement encrypted signing key storage

Implement passphrase-protected encryption/decryption for Ed25519 signing keys.

**Implementation:**
- Add write_signing_key_to_file() - encrypts and saves key to disk
- Add read_signing_key_from_file() - loads and decrypts key from disk
- Add derive_key_from_passphrase() - Argon2 KDF helper

**Cryptography:**
- Argon2 (default params) for key derivation from passphrase
- ChaCha20Poly1305 AEAD for authenticated encryption
- File format: 76 bytes [salt:16][nonce:12][ciphertext+tag:48]

**Tests:**
- Round-trip encryption/decryption
- Wrong passphrase rejection
- File corruption detection
- Invalid file size handling
- Argon2 consistency

**Dependencies:**
- argon2 = "0.5.3"
- chacha20poly1305 = "0.10.1"
- hex = "0.4" (for tests)

Closes #3

LeKarimDerradji committed:
security(crypto): add memory zeroization for passphrases

AbsoluteVirtueXI commented:
Hi, good job but, a bit too complicated:
- please remove anything related to key derivation we don't need it at this stage
- use xchacha for 192 bits nonce.
- as we need to store the nonce and the ciphered signing key a better approach is to use a file format. json is pretty good for this and is universal:
```json
{
    nonce: 123123123,
    enc_key: 13123123123
}
```
With this no need to calculate size of file and other check related to the format. If it can't deserialize then something is wrong. If i am not wrong we can even know where the serialization deserialization fail and return it to the user.
- Please check this basic example:
```rust
use crate::random;
use anyhow::{Context, Result};
use blake3;
use chacha20poly1305::{
    aead::{Aead, KeyInit},
    XChaCha20Poly1305, Key, XNonce,
};
use ed25519_dalek::{SigningKey, VerifyingKey};
use rand_core::RngCore;
use serde::{Deserialize, Serialize};
use std::{fs, path::Path};
use zeroize::Zeroize;

const NONCE_LEN: usize = 24;

/// JSON structure written to disk
#[derive(Serialize, Deserialize)]
struct SecretFile {
    nonce: String, // base64 nonce
    enc: String,   // base64 ciphertext
}

/// Derive a 32-byte symmetric key from a passphrase using BLAKE3.
/// The resulting key is ephemeral and will be zeroized after use.

fn key_from_passphrase(pass: &str) -> Key {
    let digest = blake3::hash(pass.as_bytes());
    Key::from_slice(digest.as_bytes())
}

/// Encrypt an Ed25519 signing key and write it as JSON {nonce, enc}.
pub fn write_signing_key_to_file(
    signing_key: &SigningKey,
    path: &Path,
    passphrase: &str,
) -> Result<()> {
    // Derive encryption key (32 bytes)
    let key = key_from_passphrase(passphrase);
    let cipher = XChaCha20Poly1305::new(&key);

    // Random 24-byte nonce
    let mut nonce_bytes = [0u8; NONCE_LEN];
    let mut rng = random::crypto_rng();
    rng.fill_bytes(&mut nonce_bytes);
    let nonce = XNonce::from_slice(&nonce_bytes);

    // Encrypt the signing key bytes
    let sk_bytes = signing_key.to_bytes();
    let ciphertext = cipher
        .encrypt(nonce, sk_bytes.as_ref())
        .context("encryption failed")?;

    // Zeroize key material
    let mut key_bytes = *key.as_ref();
    key_bytes.zeroize();

    // Serialize JSON
    let json = SecretFile {
        nonce: base64::encode(nonce_bytes),
        enc: base64::encode(ciphertext),
    };
    let json_str = serde_json::to_string_pretty(&json)?;
    fs::write(path, json_str)?;

    Ok(())
}

/// Read and decrypt an Ed25519 signing key from JSON file, returning both keys.
pub fn read_signing_key_from_file(
    path: &Path,
    passphrase: &str,
) -> Result<(SigningKey, VerifyingKey)> {
    // Parse JSON
    let content = fs::read_to_string(path)?;
    let parsed: SecretFile =
        serde_json::from_str(&content).context("invalid JSON format in secret file")?;

    // Decode fields
    let nonce_bytes = base64::decode(&parsed.nonce)
        .context("invalid base64 nonce in secret file")?;
    if nonce_bytes.len() != NONCE_LEN {
        anyhow::bail!("invalid nonce length");
    }
    let ciphertext =
        base64::decode(&parsed.enc).context("invalid base64 ciphertext in secret file")?;
    let nonce = XNonce::from_slice(&nonce_bytes);

    // Derive encryption key again
    let key = key_from_passphrase(passphrase);
    let cipher = XChaCha20Poly1305::new(&key);

    // Decrypt the signing key bytes
    let plaintext = cipher
        .decrypt(nonce, ciphertext.as_ref())
        .context("decryption failed (wrong passphrase or corrupted file)")?;

    // Zeroize key material
    let mut key_bytes = *key.as_ref();
    key_bytes.zeroize();

    // Convert plaintext → SigningKey
    let sk_bytes: [u8; 32] = plaintext
        .as_slice()
        .try_into()
        .context("invalid key length after decryption")?;
    let signing_key = SigningKey::from_bytes(&sk_bytes);
    let verifying_key = signing_key.verifying_key();

    Ok((signing_key, verifying_key))
}
```
need a better error handling, but i can work on it after if you want.
here is how to use it:
```rust
fn main() -> anyhow::Result<()> {
    let (vk, sk) = generate_keypair();

    let pass = "strong_password_123";
    let path = std::path::Path::new("signing_key.json");

    // write encrypted key
    write_signing_key_to_file(&sk, path, pass)?;

    // read and decrypt
    let (loaded_sk, loaded_vk) = read_signing_key_from_file(path, pass)?;

    assert_eq!(vk.to_bytes(), loaded_vk.to_bytes());
    println!("Keypair successfully recovered!");

    Ok(())
}
```

This is an example just get inspiration from. It is simple, it does what it promises, nothing more nothing less

LeKarimDerradji committed:
refactor: simplify encrypted keystore with JSON format

- Replace binary format with JSON (base64-encoded nonce/ciphertext)
- Switch from ChaCha20Poly1305 to XChaCha20Poly1305 (192-bit nonce)
- Use Blake3 for key derivation instead of Argon2
- Add zeroization of key material after encryption/decryption
- Add nonce length validation
- Update error handling to use anyhow::Result
- Update tests to match new JSON format

LeKarimDerradji commented:
Hi,

Thanks for the feedback! I've simplified the implementation as requested:

**Changes made:**
- ✅ Removed Argon2 KDF, now using Blake3 for key derivation
- ✅ Switched to XChaCha20Poly1305 with 192-bit nonce
- ✅ Implemented JSON file format with base64-encoded nonce/ciphertext
- ✅ Added key material zeroization after encryption/decryption
- ✅ Migrated to anyhow::Result for error handling
- ✅ Added nonce length validation

All tests passing (9/9). The implementation is now much simpler and follows your example structure.

Ready for review when you have time!

LeKarimDerradji committed:
docs: add SECRET_KEY_FILE constant for naming convention

- Add public constant for default encrypted key filename
- Document convention in write/read function docstrings
- Improves discoverability and standardization

LeKarimDerradji committed:
docs: fix typos and update comments for XChaCha20Poly1305

- Fix ChaCha20Poly1305 -> XChaCha20Poly1305 in comments
- Remove obsolete salt references (now using Blake3 directly)
- Fix typo: desirialize -> Deserialize
- Fix typo in test file path: decript -> decrypt
- Remove outdated Key::from_slice comment
- Improve documentation clarity
